Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift

```diff
@@ -44,153 +48,76 @@ public enum ErrorCodes: Int {
 }
+public struct AesGcmRV {
+    public var cipherText: OptionalVectorUInt8 = OptionalVectorUInt8()
+    public var errorCode: ErrorCodes = .success
+}
 
 public class AesGcm {
     public static func encrypt(
-        _ key: UnsafePointer<UInt8>,
-        keySize: UInt,
-        iv: UnsafePointer<UInt8>,
-        ivSize: UInt,
-        additionalData: UnsafePointer<UInt8>,
-        additionalDataSize: UInt,
-        plainText: UnsafePointer<UInt8>,
-        plainTextSize: UInt,
-        cipherText: UnsafeMutablePointer<UInt8>
-    ) -> ErrorCodes {
+        key: SpanConstUInt8, iv: SpanConstUInt8, ad: SpanConstUInt8, message: SpanConstUInt8,
+        desiredTagLengthInBytes: Int
+    ) -> AesGcmRV {
+        var rv = AesGcmRV()
         do {
-            if keySize > Int.max
-                || ivSize > Int.max
-                || plainTextSize > Int.max
-                || additionalDataSize > Int.max
-            {
-                return .tooBigArguments
-            }
-            if ivSize == 0 {
-                return .invalidArgument
-            }
-            let nonce = try AES.GCM.Nonce(
-                data: UnsafeBufferPointer(start: iv, count: Int(ivSize)))
-            var message = UnsafeBufferPointer(start: plainText, count: Int(plainTextSize))
-            if plainTextSize == 0 {
-                Utils.zeroArray.withUnsafeBufferPointer { ptr in
-                    message = ptr
-                }
+            if iv.size() == 0 {
+                rv.errorCode = .invalidArgument
+                return rv
             }
-            let symKey = SymmetricKey(
-                data: UnsafeBufferPointer(start: key, count: Int(keySize)))
-            let sealedBox: AES.GCM.SealedBox
-            if additionalDataSize > 0 {
-                sealedBox = try AES.GCM.seal(
-                    message,
-                    using: symKey,
-                    nonce: nonce,
-                    authenticating: UnsafeBufferPointer(
-                        start: additionalData, count: Int(additionalDataSize))
-                )
-            } else {
-                sealedBox = try AES.GCM.seal(
-                    message,
-                    using: symKey,
-                    nonce: nonce
-                )
+            let sealedBox: AES.GCM.SealedBox = try AES.GCM.seal(message, key: key, iv: iv, ad: ad)
+            var ciphertext: VectorUInt8 = Cpp.createVectorUInt8FromSpan(
+                sealedBox.ciphertext.createConstUInt8Span())
+            if desiredTagLengthInBytes > sealedBox.tag.count {
+                rv.errorCode = .invalidArgument
+                return rv
             }
-            sealedBox.ciphertext.copyBytes(to: cipherText, count: Int(plainTextSize))
-            sealedBox.tag.copyBytes(
-                to: cipherText + Int(plainTextSize), count: sealedBox.tag.count)
+            ciphertext.appendSpan(
+                span: try sealedBox.tag.createConstUInt8Span(size: desiredTagLengthInBytes))
```

Comment:
Given that sealedBox.tag and ciphertext are both Data, which is a safe type, there should be a way to append safely without the C++ conversion.

Reply:
I guess that's better. Doing that.